File: frontend/src/api/drift.ts
// SPDX-License-Identifier: AGPL-3.0-only
// Copyright (C) 2024-2026 Nexplane, Inc.

import { apiClient } from "./client";

export interface DriftEventRead {
  id: string;
  organization_id: string;
  asset_id: string;
  surface_type: string;
  drift_policy_id: string;
  baseline_state: Record<string, unknown>;
  observed_state: Record<string, unknown>;
  diff: { added: Record<string, unknown>; removed: Record<string, unknown>; changed: Record<string, { from: unknown; to: unknown }> };
  severity: "high" | "medium" | "low";
  detected_at: string;
  status: "open" | "accepted" | "attested" | "dismissed";
  shadow_cr_id: string | null;
  resolved_by: string | null;
  resolved_at: string | null;
  resolution_note: string | null;
  attested_suppress_until: string | null;
}

export interface DriftPolicyRead {
  id: string;
  organization_id: string;
  name: string;
  scope_type: string;
  scope_value: string;
  surface_types: string[];
  poll_interval_seconds: number;
  auto_created: boolean;
  enabled: boolean;
  source_cr_id: string | null;
  created_at: string;
  last_checked_at: string | null;
}

export async function listDriftEvents(params?: {
  status?: string;
  asset_id?: string;
  surface_type?: string;
  severity?: string;
}): Promise<DriftEventRead[]> {
  const query = new URLSearchParams(
    Object.fromEntries(Object.entries(params || {}).filter(([, v]) => v !== undefined)) as Record<string, string>
  );
  return apiClient.get(`/drift/events?${query}`).then((r) => r.data);
}

export async function getDriftEvent(id: string): Promise<DriftEventRead> {
  return apiClient.get(`/drift/events/${id}`).then((r) => r.data);
}

export async function acceptDriftEvent(id: string, note: string): Promise<DriftEventRead> {
  return apiClient.post(`/drift/events/${id}/accept`, { note }).then((r) => r.data);
}

export async function attestDriftEvent(id: string, note: string, snooze_days: number): Promise<DriftEventRead> {
  return apiClient.post(`/drift/events/${id}/attest`, { note, snooze_days }).then((r) => r.data);
}

export async function dismissDriftEvent(id: string): Promise<DriftEventRead> {
  return apiClient.post(`/drift/events/${id}/dismiss`).then((r) => r.data);
}

export async function getAssetDrift(assetId: string): Promise<{
  asset_id: string;
  resource_states: Array<{ surface_type: string; state: Record<string, unknown>; captured_at: string; source: string }>;
  open_events: DriftEventRead[];
}> {
  return apiClient.get(`/assets/${assetId}/drift`).then((r) => r.data);
}

File: backend/app/schemas/drift.py
# SPDX-License-Identifier: AGPL-3.0-only
# Copyright (C) 2024-2026 Nexplane, Inc.

import uuid
from datetime import datetime
from typing import Optional
from pydantic import BaseModel


class ResourceStateRead(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    organization_id: uuid.UUID
    asset_id: uuid.UUID
    surface_type: str
    state: dict
    captured_at: datetime
    source: str
    source_cr_id: Optional[uuid.UUID] = None
    accepted_by: Optional[uuid.UUID] = None
    accepted_at: Optional[datetime] = None
    acceptance_note: Optional[str] = None


class DriftPolicyCreate(BaseModel):
    name: str
    scope_type: str
    scope_value: str
    surface_types: list[str]
    poll_interval_seconds: int = 3600
    enabled: bool = True


class DriftPolicyRead(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    organization_id: uuid.UUID
    name: str
    scope_type: str
    scope_value: str
    surface_types: list[str]
    poll_interval_seconds: int
    auto_created: bool
    enabled: bool
    source_cr_id: Optional[uuid.UUID] = None
    created_by: Optional[uuid.UUID] = None
    created_at: datetime
    last_checked_at: Optional[datetime] = None


class DriftPolicyUpdate(BaseModel):
    name: Optional[str] = None
    poll_interval_seconds: Optional[int] = None
    enabled: Optional[bool] = None
    surface_types: Optional[list[str]] = None


class DriftEventRead(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    organization_id: uuid.UUID
    asset_id: uuid.UUID
    surface_type: str
    drift_policy_id: uuid.UUID
    baseline_state: dict
    observed_state: dict
    diff: dict
    severity: str
    detected_at: datetime
    status: str
    shadow_cr_id: Optional[uuid.UUID] = None
    resolved_by: Optional[uuid.UUID] = None
    resolved_at: Optional[datetime] = None
    resolution_note: Optional[str] = None
    attested_suppress_until: Optional[datetime] = None


class DriftEventAcceptBody(BaseModel):
    note: str = ""


class DriftEventAttestBody(BaseModel):
    note: str = ""
    snooze_days: int = 7


class AssetDriftSummary(BaseModel):
    asset_id: uuid.UUID
    resource_states: list[ResourceStateRead]
    open_events: list[DriftEventRead]
